### ISSUE TITLE
Reference policy ablations

### ISSUE THREAD

**Issue by yesiam-png** [RLHFlow/Online-RLHF#18]
Dear authors, 
1. I noticed that the reference policy is fixed as the initial policy, instead of updating as the last iter's policy. May I know the reason for it and have you tried updating the reference policy? 
2. The initial policy is "RLHFlow/LLaMA3-SFT", not "meta-llama/Meta-Llama-3-8B-Instruct", right?

**Comment by hendrydong** [comment]
From the theoretical perspective [1], the target should be fixed, so that the ref model should be fixed. I tried some experiments of changing the policy. It may perform well with proper tuning, but the whole system is more unstable.

Yes. 


[1] iterative preference learning from human feedback: bridging theory and practice for rlhf under kl-constraint

**Comment by WeiXiongUST** [comment]
if you use a fixed reference policy, you always search your policy within a kl ball of the sft model. Instead, if you update the reference policy per-iteration, it is like a mirror descent and essentially you are optimizing the reward function without the kl regularization. 

The intuition is that while you constraint the model not to move too far away from the reference model at each iteration (because you have a kl in dpo), the cumulative update can make the model move far away.

You can see that the resulting model of our project maintains a very competitive performance on the academic benchmarks. You can definitely update the reference policy to get a better in-domain performance like mt-bench, alpha-eval score. It is just like a trade-off.

**Comment by yesiam-png** [comment]
Thank you all for your response! That makes sense. Since I was unable to run gen_hf.py due to port issues of my server, I used something similar to get_hf.py using multiprocessing. I have double-checked the function calling, hyperparameters, and outputs, they all seem to align with your choices. But I ended up getting an iter-3 model with LC Alpaca score 22 (I have also double-checked the chat template). Could you confirm the hyperparameters, e.g., bs=8(gpus)*1(bs_per_device)*16(accumulate)=128, lr=5e-7 (in your paper it seems to be 5e-6?) are correct to reproduce RLHFlow/LLaMA3-iterative-DPO-final?

**Comment by WeiXiongUST** [comment]
The parameter is supposed to be 5e-7. I have heard several friends that their reproduced results are worse with ~25 alpaca eval score but also see some research papers reproduce with even better results... I am not sure what is the key difference here. One thing you can try is to update the reference policy to get a better in-domain performance (alpaca eval, mt-bench...) . 

I will have a paper hopefully released this month and will have a more comprehensive discussion about their differences. Then, I will update the code base accordingly.

**Comment by yesiam-png** [comment]
Dear authors, I have tried updating the reference policy but it still gives bad results (~22 on LC Alpaca). May I know if you could get LC alpaca >30 using the same code from this repo? Is it possible to share your recipe for reproducing RLHFlow/LLaMA3-iterative-DPO-final?

**Comment by hendrydong** [comment]
I think it should be some hyper parameter tuning issues.

Have you seen the length of your response when testing the Alpaca?

**Comment by yesiam-png** [comment]
The length is 2033, which I think is within a normal range. Could you share the same training recipe for reproduction? I don't think iDPO itself is hard to implement, but the hyperparameters and tricks will make this repo valuable.

**Comment by WeiXiongUST** [comment]
Actually, I have checked that the hyper-parameters are exactly what we use to make the model in the report and we do not use specialized tricks (except the rejection sampling mentioned in the paper)....

Maybe looking into the training record and double checking the dataset used for each round of experiments can help. Also, you may also want to make sure that the GPT-4 model and config are the same for the alpaca eval benchmark evaluation.
